// square/dist/serialization/types/CatalogProductSet.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CatalogProductSet = void 0;
const core = __importStar(require("../../core"));
exports.CatalogProductSet = core.serialization.object({
    name: core.serialization.string().optionalNullable(),
    productIdsAny: core.serialization.property("product_ids_any", core.serialization.list(core.serialization.string()).optionalNullable()),
    productIdsAll: core.serialization.property("product_ids_all", core.serialization.list(core.serialization.string()).optionalNullable()),
    quantityExact: core.serialization.property("quantity_exact", core.serialization.bigint().optionalNullable()),
    quantityMin: core.serialization.property("quantity_min", core.serialization.bigint().optionalNullable()),
    quantityMax: core.serialization.property("quantity_max", core.serialization.bigint().optionalNullable()),
    allProducts: core.serialization.property("all_products", core.serialization.boolean().optionalNullable()),
});
