// square/dist/serialization/types/GiftCard.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GiftCard = void 0;
const core = __importStar(require("../../core"));
const GiftCardType_1 = require("./GiftCardType");
const GiftCardGanSource_1 = require("./GiftCardGanSource");
const GiftCardStatus_1 = require("./GiftCardStatus");
const Money_1 = require("./Money");
exports.GiftCard = core.serialization.object({
    id: core.serialization.string().optional(),
    type: GiftCardType_1.GiftCardType,
    ganSource: core.serialization.property("gan_source", GiftCardGanSource_1.GiftCardGanSource.optional()),
    state: GiftCardStatus_1.GiftCardStatus.optional(),
    balanceMoney: core.serialization.property("balance_money", Money_1.Money.optional()),
    gan: core.serialization.string().optionalNullable(),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    customerIds: core.serialization.property("customer_ids", core.serialization.list(core.serialization.string()).optional()),
});
