// square/dist/serialization/types/InvoicePaymentRequest.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvoicePaymentRequest = void 0;
const core = __importStar(require("../../core"));
const InvoiceRequestMethod_1 = require("./InvoiceRequestMethod");
const InvoiceRequestType_1 = require("./InvoiceRequestType");
const Money_1 = require("./Money");
const InvoiceAutomaticPaymentSource_1 = require("./InvoiceAutomaticPaymentSource");
const InvoicePaymentReminder_1 = require("./InvoicePaymentReminder");
exports.InvoicePaymentRequest = core.serialization.object({
    uid: core.serialization.string().optionalNullable(),
    requestMethod: core.serialization.property("request_method", InvoiceRequestMethod_1.InvoiceRequestMethod.optional()),
    requestType: core.serialization.property("request_type", InvoiceRequestType_1.InvoiceRequestType.optional()),
    dueDate: core.serialization.property("due_date", core.serialization.string().optionalNullable()),
    fixedAmountRequestedMoney: core.serialization.property("fixed_amount_requested_money", Money_1.Money.optional()),
    percentageRequested: core.serialization.property("percentage_requested", core.serialization.string().optionalNullable()),
    tippingEnabled: core.serialization.property("tipping_enabled", core.serialization.boolean().optionalNullable()),
    automaticPaymentSource: core.serialization.property("automatic_payment_source", InvoiceAutomaticPaymentSource_1.InvoiceAutomaticPaymentSource.optional()),
    cardId: core.serialization.property("card_id", core.serialization.string().optionalNullable()),
    reminders: core.serialization.list(InvoicePaymentReminder_1.InvoicePaymentReminder).optionalNullable(),
    computedAmountMoney: core.serialization.property("computed_amount_money", Money_1.Money.optional()),
    totalCompletedAmountMoney: core.serialization.property("total_completed_amount_money", Money_1.Money.optional()),
    roundingAdjustmentIncludedMoney: core.serialization.property("rounding_adjustment_included_money", Money_1.Money.optional()),
});
