// square/dist/serialization/types/LoyaltyPromotion.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoyaltyPromotion = void 0;
const core = __importStar(require("../../core"));
const LoyaltyPromotionIncentive_1 = require("./LoyaltyPromotionIncentive");
const LoyaltyPromotionAvailableTimeData_1 = require("./LoyaltyPromotionAvailableTimeData");
const LoyaltyPromotionTriggerLimit_1 = require("./LoyaltyPromotionTriggerLimit");
const LoyaltyPromotionStatus_1 = require("./LoyaltyPromotionStatus");
const Money_1 = require("./Money");
exports.LoyaltyPromotion = core.serialization.object({
    id: core.serialization.string().optional(),
    name: core.serialization.string(),
    incentive: LoyaltyPromotionIncentive_1.LoyaltyPromotionIncentive,
    availableTime: core.serialization.property("available_time", LoyaltyPromotionAvailableTimeData_1.LoyaltyPromotionAvailableTimeData),
    triggerLimit: core.serialization.property("trigger_limit", LoyaltyPromotionTriggerLimit_1.LoyaltyPromotionTriggerLimit.optional()),
    status: LoyaltyPromotionStatus_1.LoyaltyPromotionStatus.optional(),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    canceledAt: core.serialization.property("canceled_at", core.serialization.string().optional()),
    updatedAt: core.serialization.property("updated_at", core.serialization.string().optional()),
    loyaltyProgramId: core.serialization.property("loyalty_program_id", core.serialization.string().optional()),
    minimumSpendAmountMoney: core.serialization.property("minimum_spend_amount_money", Money_1.Money.optional()),
    qualifyingItemVariationIds: core.serialization.property("qualifying_item_variation_ids", core.serialization.list(core.serialization.string()).optionalNullable()),
    qualifyingCategoryIds: core.serialization.property("qualifying_category_ids", core.serialization.list(core.serialization.string()).optionalNullable()),
});
