// square/dist/serialization/types/Payout.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payout = void 0;
const core = __importStar(require("../../core"));
const PayoutStatus_1 = require("./PayoutStatus");
const Money_1 = require("./Money");
const Destination_1 = require("./Destination");
const PayoutType_1 = require("./PayoutType");
const PayoutFee_1 = require("./PayoutFee");
exports.Payout = core.serialization.object({
    id: core.serialization.string(),
    status: PayoutStatus_1.PayoutStatus.optional(),
    locationId: core.serialization.property("location_id", core.serialization.string()),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    updatedAt: core.serialization.property("updated_at", core.serialization.string().optional()),
    amountMoney: core.serialization.property("amount_money", Money_1.Money.optional()),
    destination: Destination_1.Destination.optional(),
    version: core.serialization.number().optional(),
    type: PayoutType_1.PayoutType.optional(),
    payoutFee: core.serialization.property("payout_fee", core.serialization.list(PayoutFee_1.PayoutFee).optionalNullable()),
    arrivalDate: core.serialization.property("arrival_date", core.serialization.string().optionalNullable()),
    endToEndId: core.serialization.property("end_to_end_id", core.serialization.string().optionalNullable()),
});
