// square/dist/serialization/types/Tender.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tender = void 0;
const core = __importStar(require("../../core"));
const Money_1 = require("./Money");
const TenderType_1 = require("./TenderType");
const TenderCardDetails_1 = require("./TenderCardDetails");
const TenderCashDetails_1 = require("./TenderCashDetails");
const TenderBankAccountDetails_1 = require("./TenderBankAccountDetails");
const TenderBuyNowPayLaterDetails_1 = require("./TenderBuyNowPayLaterDetails");
const TenderSquareAccountDetails_1 = require("./TenderSquareAccountDetails");
const AdditionalRecipient_1 = require("./AdditionalRecipient");
exports.Tender = core.serialization.object({
    id: core.serialization.string().optional(),
    locationId: core.serialization.property("location_id", core.serialization.string().optionalNullable()),
    transactionId: core.serialization.property("transaction_id", core.serialization.string().optionalNullable()),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    note: core.serialization.string().optionalNullable(),
    amountMoney: core.serialization.property("amount_money", Money_1.Money.optional()),
    tipMoney: core.serialization.property("tip_money", Money_1.Money.optional()),
    processingFeeMoney: core.serialization.property("processing_fee_money", Money_1.Money.optional()),
    customerId: core.serialization.property("customer_id", core.serialization.string().optionalNullable()),
    type: TenderType_1.TenderType,
    cardDetails: core.serialization.property("card_details", TenderCardDetails_1.TenderCardDetails.optional()),
    cashDetails: core.serialization.property("cash_details", TenderCashDetails_1.TenderCashDetails.optional()),
    bankAccountDetails: core.serialization.property("bank_account_details", TenderBankAccountDetails_1.TenderBankAccountDetails.optional()),
    buyNowPayLaterDetails: core.serialization.property("buy_now_pay_later_details", TenderBuyNowPayLaterDetails_1.TenderBuyNowPayLaterDetails.optional()),
    squareAccountDetails: core.serialization.property("square_account_details", TenderSquareAccountDetails_1.TenderSquareAccountDetails.optional()),
    additionalRecipients: core.serialization.property("additional_recipients", core.serialization.list(AdditionalRecipient_1.AdditionalRecipient).optionalNullable()),
    paymentId: core.serialization.property("payment_id", core.serialization.string().optionalNullable()),
});
