// square/dist/serialization/types/Transaction.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const core = __importStar(require("../../core"));
const Tender_1 = require("./Tender");
const Refund_1 = require("./Refund");
const TransactionProduct_1 = require("./TransactionProduct");
const Address_1 = require("./Address");
exports.Transaction = core.serialization.object({
    id: core.serialization.string().optional(),
    locationId: core.serialization.property("location_id", core.serialization.string().optionalNullable()),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    tenders: core.serialization.list(Tender_1.Tender).optionalNullable(),
    refunds: core.serialization.list(Refund_1.Refund).optionalNullable(),
    referenceId: core.serialization.property("reference_id", core.serialization.string().optionalNullable()),
    product: TransactionProduct_1.TransactionProduct.optional(),
    clientId: core.serialization.property("client_id", core.serialization.string().optionalNullable()),
    shippingAddress: core.serialization.property("shipping_address", Address_1.Address.optional()),
    orderId: core.serialization.property("order_id", core.serialization.string().optionalNullable()),
});
