// square/dist/api/resources/labor/client/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Labor = void 0;
const Client_1 = require("../resources/breakTypes/client/Client");
const Client_2 = require("../resources/employeeWages/client/Client");
const Client_3 = require("../resources/shifts/client/Client");
const Client_4 = require("../resources/teamMemberWages/client/Client");
const Client_5 = require("../resources/workweekConfigs/client/Client");
class Labor {
    constructor(_options = {}) {
        this._options = _options;
    }
    get breakTypes() {
        var _a;
        return ((_a = this._breakTypes) !== null && _a !== void 0 ? _a : (this._breakTypes = new Client_1.BreakTypes(this._options)));
    }
    get employeeWages() {
        var _a;
        return ((_a = this._employeeWages) !== null && _a !== void 0 ? _a : (this._employeeWages = new Client_2.EmployeeWages(this._options)));
    }
    get shifts() {
        var _a;
        return ((_a = this._shifts) !== null && _a !== void 0 ? _a : (this._shifts = new Client_3.Shifts(this._options)));
    }
    get teamMemberWages() {
        var _a;
        return ((_a = this._teamMemberWages) !== null && _a !== void 0 ? _a : (this._teamMemberWages = new Client_4.TeamMemberWages(this._options)));
    }
    get workweekConfigs() {
        var _a;
        return ((_a = this._workweekConfigs) !== null && _a !== void 0 ? _a : (this._workweekConfigs = new Client_5.WorkweekConfigs(this._options)));
    }
}
exports.Labor = Labor;
