// square/serialization/types/LoyaltyEvent.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoyaltyEvent = void 0;
const core = __importStar(require("../../core"));
const LoyaltyEventType_1 = require("./LoyaltyEventType");
const LoyaltyEventAccumulatePoints_1 = require("./LoyaltyEventAccumulatePoints");
const LoyaltyEventCreateReward_1 = require("./LoyaltyEventCreateReward");
const LoyaltyEventRedeemReward_1 = require("./LoyaltyEventRedeemReward");
const LoyaltyEventDeleteReward_1 = require("./LoyaltyEventDeleteReward");
const LoyaltyEventAdjustPoints_1 = require("./LoyaltyEventAdjustPoints");
const LoyaltyEventSource_1 = require("./LoyaltyEventSource");
const LoyaltyEventExpirePoints_1 = require("./LoyaltyEventExpirePoints");
const LoyaltyEventOther_1 = require("./LoyaltyEventOther");
const LoyaltyEventAccumulatePromotionPoints_1 = require("./LoyaltyEventAccumulatePromotionPoints");
exports.LoyaltyEvent = core.serialization.object({
    id: core.serialization.string(),
    type: LoyaltyEventType_1.LoyaltyEventType,
    createdAt: core.serialization.property("created_at", core.serialization.string()),
    accumulatePoints: core.serialization.property("accumulate_points", LoyaltyEventAccumulatePoints_1.LoyaltyEventAccumulatePoints.optional()),
    createReward: core.serialization.property("create_reward", LoyaltyEventCreateReward_1.LoyaltyEventCreateReward.optional()),
    redeemReward: core.serialization.property("redeem_reward", LoyaltyEventRedeemReward_1.LoyaltyEventRedeemReward.optional()),
    deleteReward: core.serialization.property("delete_reward", LoyaltyEventDeleteReward_1.LoyaltyEventDeleteReward.optional()),
    adjustPoints: core.serialization.property("adjust_points", LoyaltyEventAdjustPoints_1.LoyaltyEventAdjustPoints.optional()),
    loyaltyAccountId: core.serialization.property("loyalty_account_id", core.serialization.string()),
    locationId: core.serialization.property("location_id", core.serialization.string().optional()),
    source: LoyaltyEventSource_1.LoyaltyEventSource,
    expirePoints: core.serialization.property("expire_points", LoyaltyEventExpirePoints_1.LoyaltyEventExpirePoints.optional()),
    otherEvent: core.serialization.property("other_event", LoyaltyEventOther_1.LoyaltyEventOther.optional()),
    accumulatePromotionPoints: core.serialization.property("accumulate_promotion_points", LoyaltyEventAccumulatePromotionPoints_1.LoyaltyEventAccumulatePromotionPoints.optional()),
});
