// square/serialization/types/LoyaltyProgram.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoyaltyProgram = void 0;
const core = __importStar(require("../../core"));
const LoyaltyProgramStatus_1 = require("./LoyaltyProgramStatus");
const LoyaltyProgramRewardTier_1 = require("./LoyaltyProgramRewardTier");
const LoyaltyProgramExpirationPolicy_1 = require("./LoyaltyProgramExpirationPolicy");
const LoyaltyProgramTerminology_1 = require("./LoyaltyProgramTerminology");
const LoyaltyProgramAccrualRule_1 = require("./LoyaltyProgramAccrualRule");
exports.LoyaltyProgram = core.serialization.object({
    id: core.serialization.string().optional(),
    status: LoyaltyProgramStatus_1.LoyaltyProgramStatus.optional(),
    rewardTiers: core.serialization.property("reward_tiers", core.serialization.list(LoyaltyProgramRewardTier_1.LoyaltyProgramRewardTier).optionalNullable()),
    expirationPolicy: core.serialization.property("expiration_policy", LoyaltyProgramExpirationPolicy_1.LoyaltyProgramExpirationPolicy.optional()),
    terminology: LoyaltyProgramTerminology_1.LoyaltyProgramTerminology.optional(),
    locationIds: core.serialization.property("location_ids", core.serialization.list(core.serialization.string()).optionalNullable()),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    updatedAt: core.serialization.property("updated_at", core.serialization.string().optional()),
    accrualRules: core.serialization.property("accrual_rules", core.serialization.list(LoyaltyProgramAccrualRule_1.LoyaltyProgramAccrualRule).optionalNullable()),
});
