// square/serialization/types/LoyaltyEventFilter.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoyaltyEventFilter = void 0;
const core = __importStar(require("../../core"));
const LoyaltyEventLoyaltyAccountFilter_1 = require("./LoyaltyEventLoyaltyAccountFilter");
const LoyaltyEventTypeFilter_1 = require("./LoyaltyEventTypeFilter");
const LoyaltyEventDateTimeFilter_1 = require("./LoyaltyEventDateTimeFilter");
const LoyaltyEventLocationFilter_1 = require("./LoyaltyEventLocationFilter");
const LoyaltyEventOrderFilter_1 = require("./LoyaltyEventOrderFilter");
exports.LoyaltyEventFilter = core.serialization.object({
    loyaltyAccountFilter: core.serialization.property("loyalty_account_filter", LoyaltyEventLoyaltyAccountFilter_1.LoyaltyEventLoyaltyAccountFilter.optional()),
    typeFilter: core.serialization.property("type_filter", LoyaltyEventTypeFilter_1.LoyaltyEventTypeFilter.optional()),
    dateTimeFilter: core.serialization.property("date_time_filter", LoyaltyEventDateTimeFilter_1.LoyaltyEventDateTimeFilter.optional()),
    locationFilter: core.serialization.property("location_filter", LoyaltyEventLocationFilter_1.LoyaltyEventLocationFilter.optional()),
    orderFilter: core.serialization.property("order_filter", LoyaltyEventOrderFilter_1.LoyaltyEventOrderFilter.optional()),
});
