// square/serialization/types/MeasurementUnit.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeasurementUnit = void 0;
const core = __importStar(require("../../core"));
const MeasurementUnitCustom_1 = require("./MeasurementUnitCustom");
const MeasurementUnitArea_1 = require("./MeasurementUnitArea");
const MeasurementUnitLength_1 = require("./MeasurementUnitLength");
const MeasurementUnitVolume_1 = require("./MeasurementUnitVolume");
const MeasurementUnitWeight_1 = require("./MeasurementUnitWeight");
const MeasurementUnitGeneric_1 = require("./MeasurementUnitGeneric");
const MeasurementUnitTime_1 = require("./MeasurementUnitTime");
const MeasurementUnitUnitType_1 = require("./MeasurementUnitUnitType");
exports.MeasurementUnit = core.serialization.object({
    customUnit: core.serialization.property("custom_unit", MeasurementUnitCustom_1.MeasurementUnitCustom.optional()),
    areaUnit: core.serialization.property("area_unit", MeasurementUnitArea_1.MeasurementUnitArea.optional()),
    lengthUnit: core.serialization.property("length_unit", MeasurementUnitLength_1.MeasurementUnitLength.optional()),
    volumeUnit: core.serialization.property("volume_unit", MeasurementUnitVolume_1.MeasurementUnitVolume.optional()),
    weightUnit: core.serialization.property("weight_unit", MeasurementUnitWeight_1.MeasurementUnitWeight.optional()),
    genericUnit: core.serialization.property("generic_unit", MeasurementUnitGeneric_1.MeasurementUnitGeneric.optional()),
    timeUnit: core.serialization.property("time_unit", MeasurementUnitTime_1.MeasurementUnitTime.optional()),
    type: MeasurementUnitUnitType_1.MeasurementUnitUnitType.optional(),
});
