// square/serialization/types/ObtainTokenResponse.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObtainTokenResponse = void 0;
const core = __importStar(require("../../core"));
const Error_1 = require("./Error_");
exports.ObtainTokenResponse = core.serialization.object({
    accessToken: core.serialization.property("access_token", core.serialization.string().optional()),
    tokenType: core.serialization.property("token_type", core.serialization.string().optional()),
    expiresAt: core.serialization.property("expires_at", core.serialization.string().optional()),
    merchantId: core.serialization.property("merchant_id", core.serialization.string().optional()),
    subscriptionId: core.serialization.property("subscription_id", core.serialization.string().optional()),
    planId: core.serialization.property("plan_id", core.serialization.string().optional()),
    idToken: core.serialization.property("id_token", core.serialization.string().optional()),
    refreshToken: core.serialization.property("refresh_token", core.serialization.string().optional()),
    shortLived: core.serialization.property("short_lived", core.serialization.boolean().optional()),
    errors: core.serialization.list(Error_1.Error_).optional(),
    refreshTokenExpiresAt: core.serialization.property("refresh_token_expires_at", core.serialization.string().optional()),
});
