// square/serialization/types/OrderReturnLineItemModifier.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderReturnLineItemModifier = void 0;
const core = __importStar(require("../../core"));
const Money_1 = require("./Money");
exports.OrderReturnLineItemModifier = core.serialization.object({
    uid: core.serialization.string().optionalNullable(),
    sourceModifierUid: core.serialization.property("source_modifier_uid", core.serialization.string().optionalNullable()),
    catalogObjectId: core.serialization.property("catalog_object_id", core.serialization.string().optionalNullable()),
    catalogVersion: core.serialization.property("catalog_version", core.serialization.bigint().optionalNullable()),
    name: core.serialization.string().optionalNullable(),
    basePriceMoney: core.serialization.property("base_price_money", Money_1.Money.optional()),
    totalPriceMoney: core.serialization.property("total_price_money", Money_1.Money.optional()),
    quantity: core.serialization.string().optionalNullable(),
});
