// square/serialization/types/OrderReturn.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderReturn = void 0;
const core = __importStar(require("../../core"));
const OrderReturnLineItem_1 = require("./OrderReturnLineItem");
const OrderReturnServiceCharge_1 = require("./OrderReturnServiceCharge");
const OrderReturnTax_1 = require("./OrderReturnTax");
const OrderReturnDiscount_1 = require("./OrderReturnDiscount");
const OrderReturnTip_1 = require("./OrderReturnTip");
const OrderRoundingAdjustment_1 = require("./OrderRoundingAdjustment");
const OrderMoneyAmounts_1 = require("./OrderMoneyAmounts");
exports.OrderReturn = core.serialization.object({
    uid: core.serialization.string().optionalNullable(),
    sourceOrderId: core.serialization.property("source_order_id", core.serialization.string().optionalNullable()),
    returnLineItems: core.serialization.property("return_line_items", core.serialization.list(OrderReturnLineItem_1.OrderReturnLineItem).optionalNullable()),
    returnServiceCharges: core.serialization.property("return_service_charges", core.serialization.list(OrderReturnServiceCharge_1.OrderReturnServiceCharge).optionalNullable()),
    returnTaxes: core.serialization.property("return_taxes", core.serialization.list(OrderReturnTax_1.OrderReturnTax).optional()),
    returnDiscounts: core.serialization.property("return_discounts", core.serialization.list(OrderReturnDiscount_1.OrderReturnDiscount).optional()),
    returnTips: core.serialization.property("return_tips", core.serialization.list(OrderReturnTip_1.OrderReturnTip).optionalNullable()),
    roundingAdjustment: core.serialization.property("rounding_adjustment", OrderRoundingAdjustment_1.OrderRoundingAdjustment.optional()),
    returnAmounts: core.serialization.property("return_amounts", OrderMoneyAmounts_1.OrderMoneyAmounts.optional()),
});
