// square/serialization/types/OrderLineItemTax.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderLineItemTax = void 0;
const core = __importStar(require("../../core"));
const OrderLineItemTaxType_1 = require("./OrderLineItemTaxType");
const Money_1 = require("./Money");
const OrderLineItemTaxScope_1 = require("./OrderLineItemTaxScope");
exports.OrderLineItemTax = core.serialization.object({
    uid: core.serialization.string().optionalNullable(),
    catalogObjectId: core.serialization.property("catalog_object_id", core.serialization.string().optionalNullable()),
    catalogVersion: core.serialization.property("catalog_version", core.serialization.bigint().optionalNullable()),
    name: core.serialization.string().optionalNullable(),
    type: OrderLineItemTaxType_1.OrderLineItemTaxType.optional(),
    percentage: core.serialization.string().optionalNullable(),
    metadata: core.serialization
        .record(core.serialization.string(), core.serialization.string().optionalNullable())
        .optionalNullable(),
    appliedMoney: core.serialization.property("applied_money", Money_1.Money.optional()),
    scope: OrderLineItemTaxScope_1.OrderLineItemTaxScope.optional(),
    autoApplied: core.serialization.property("auto_applied", core.serialization.boolean().optional()),
});
