// square/serialization/types/OrderLineItem.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderLineItem = void 0;
const core = __importStar(require("../../core"));
const OrderQuantityUnit_1 = require("./OrderQuantityUnit");
const OrderLineItemItemType_1 = require("./OrderLineItemItemType");
const OrderLineItemModifier_1 = require("./OrderLineItemModifier");
const OrderLineItemAppliedTax_1 = require("./OrderLineItemAppliedTax");
const OrderLineItemAppliedDiscount_1 = require("./OrderLineItemAppliedDiscount");
const OrderLineItemAppliedServiceCharge_1 = require("./OrderLineItemAppliedServiceCharge");
const Money_1 = require("./Money");
const OrderLineItemPricingBlocklists_1 = require("./OrderLineItemPricingBlocklists");
exports.OrderLineItem = core.serialization.object({
    uid: core.serialization.string().optionalNullable(),
    name: core.serialization.string().optionalNullable(),
    quantity: core.serialization.string(),
    quantityUnit: core.serialization.property("quantity_unit", OrderQuantityUnit_1.OrderQuantityUnit.optional()),
    note: core.serialization.string().optionalNullable(),
    catalogObjectId: core.serialization.property("catalog_object_id", core.serialization.string().optionalNullable()),
    catalogVersion: core.serialization.property("catalog_version", core.serialization.bigint().optionalNullable()),
    variationName: core.serialization.property("variation_name", core.serialization.string().optionalNullable()),
    itemType: core.serialization.property("item_type", OrderLineItemItemType_1.OrderLineItemItemType.optional()),
    metadata: core.serialization
        .record(core.serialization.string(), core.serialization.string().optionalNullable())
        .optionalNullable(),
    modifiers: core.serialization.list(OrderLineItemModifier_1.OrderLineItemModifier).optionalNullable(),
    appliedTaxes: core.serialization.property("applied_taxes", core.serialization.list(OrderLineItemAppliedTax_1.OrderLineItemAppliedTax).optionalNullable()),
    appliedDiscounts: core.serialization.property("applied_discounts", core.serialization.list(OrderLineItemAppliedDiscount_1.OrderLineItemAppliedDiscount).optionalNullable()),
    appliedServiceCharges: core.serialization.property("applied_service_charges", core.serialization.list(OrderLineItemAppliedServiceCharge_1.OrderLineItemAppliedServiceCharge).optionalNullable()),
    basePriceMoney: core.serialization.property("base_price_money", Money_1.Money.optional()),
    variationTotalPriceMoney: core.serialization.property("variation_total_price_money", Money_1.Money.optional()),
    grossSalesMoney: core.serialization.property("gross_sales_money", Money_1.Money.optional()),
    totalTaxMoney: core.serialization.property("total_tax_money", Money_1.Money.optional()),
    totalDiscountMoney: core.serialization.property("total_discount_money", Money_1.Money.optional()),
    totalMoney: core.serialization.property("total_money", Money_1.Money.optional()),
    pricingBlocklists: core.serialization.property("pricing_blocklists", OrderLineItemPricingBlocklists_1.OrderLineItemPricingBlocklists.optional()),
    totalServiceChargeMoney: core.serialization.property("total_service_charge_money", Money_1.Money.optional()),
});
