// square/serialization/types/Payment.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Payment = void 0;
const core = __importStar(require("../../core"));
const Money_1 = require("./Money");
const ProcessingFee_1 = require("./ProcessingFee");
const CardPaymentDetails_1 = require("./CardPaymentDetails");
const CashPaymentDetails_1 = require("./CashPaymentDetails");
const BankAccountPaymentDetails_1 = require("./BankAccountPaymentDetails");
const ExternalPaymentDetails_1 = require("./ExternalPaymentDetails");
const DigitalWalletDetails_1 = require("./DigitalWalletDetails");
const BuyNowPayLaterDetails_1 = require("./BuyNowPayLaterDetails");
const SquareAccountDetails_1 = require("./SquareAccountDetails");
const RiskEvaluation_1 = require("./RiskEvaluation");
const Address_1 = require("./Address");
const DeviceDetails_1 = require("./DeviceDetails");
const ApplicationDetails_1 = require("./ApplicationDetails");
const OfflinePaymentDetails_1 = require("./OfflinePaymentDetails");
exports.Payment = core.serialization.object({
    id: core.serialization.string().optional(),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    updatedAt: core.serialization.property("updated_at", core.serialization.string().optional()),
    amountMoney: core.serialization.property("amount_money", Money_1.Money.optional()),
    tipMoney: core.serialization.property("tip_money", Money_1.Money.optional()),
    totalMoney: core.serialization.property("total_money", Money_1.Money.optional()),
    appFeeMoney: core.serialization.property("app_fee_money", Money_1.Money.optional()),
    approvedMoney: core.serialization.property("approved_money", Money_1.Money.optional()),
    processingFee: core.serialization.property("processing_fee", core.serialization.list(ProcessingFee_1.ProcessingFee).optional()),
    refundedMoney: core.serialization.property("refunded_money", Money_1.Money.optional()),
    status: core.serialization.string().optional(),
    delayDuration: core.serialization.property("delay_duration", core.serialization.string().optional()),
    delayAction: core.serialization.property("delay_action", core.serialization.string().optionalNullable()),
    delayedUntil: core.serialization.property("delayed_until", core.serialization.string().optional()),
    sourceType: core.serialization.property("source_type", core.serialization.string().optional()),
    cardDetails: core.serialization.property("card_details", CardPaymentDetails_1.CardPaymentDetails.optional()),
    cashDetails: core.serialization.property("cash_details", CashPaymentDetails_1.CashPaymentDetails.optional()),
    bankAccountDetails: core.serialization.property("bank_account_details", BankAccountPaymentDetails_1.BankAccountPaymentDetails.optional()),
    externalDetails: core.serialization.property("external_details", ExternalPaymentDetails_1.ExternalPaymentDetails.optional()),
    walletDetails: core.serialization.property("wallet_details", DigitalWalletDetails_1.DigitalWalletDetails.optional()),
    buyNowPayLaterDetails: core.serialization.property("buy_now_pay_later_details", BuyNowPayLaterDetails_1.BuyNowPayLaterDetails.optional()),
    squareAccountDetails: core.serialization.property("square_account_details", SquareAccountDetails_1.SquareAccountDetails.optional()),
    locationId: core.serialization.property("location_id", core.serialization.string().optional()),
    orderId: core.serialization.property("order_id", core.serialization.string().optional()),
    referenceId: core.serialization.property("reference_id", core.serialization.string().optional()),
    customerId: core.serialization.property("customer_id", core.serialization.string().optional()),
    employeeId: core.serialization.property("employee_id", core.serialization.string().optional()),
    teamMemberId: core.serialization.property("team_member_id", core.serialization.string().optionalNullable()),
    refundIds: core.serialization.property("refund_ids", core.serialization.list(core.serialization.string()).optional()),
    riskEvaluation: core.serialization.property("risk_evaluation", RiskEvaluation_1.RiskEvaluation.optional()),
    terminalCheckoutId: core.serialization.property("terminal_checkout_id", core.serialization.string().optional()),
    buyerEmailAddress: core.serialization.property("buyer_email_address", core.serialization.string().optional()),
    billingAddress: core.serialization.property("billing_address", Address_1.Address.optional()),
    shippingAddress: core.serialization.property("shipping_address", Address_1.Address.optional()),
    note: core.serialization.string().optional(),
    statementDescriptionIdentifier: core.serialization.property("statement_description_identifier", core.serialization.string().optional()),
    capabilities: core.serialization.list(core.serialization.string()).optional(),
    receiptNumber: core.serialization.property("receipt_number", core.serialization.string().optional()),
    receiptUrl: core.serialization.property("receipt_url", core.serialization.string().optional()),
    deviceDetails: core.serialization.property("device_details", DeviceDetails_1.DeviceDetails.optional()),
    applicationDetails: core.serialization.property("application_details", ApplicationDetails_1.ApplicationDetails.optional()),
    isOfflinePayment: core.serialization.property("is_offline_payment", core.serialization.boolean().optional()),
    offlinePaymentDetails: core.serialization.property("offline_payment_details", OfflinePaymentDetails_1.OfflinePaymentDetails.optional()),
    versionToken: core.serialization.property("version_token", core.serialization.string().optionalNullable()),
});
