// square/serialization/types/Subscription.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Subscription = void 0;
const core = __importStar(require("../../core"));
const SubscriptionStatus_1 = require("./SubscriptionStatus");
const Money_1 = require("./Money");
const SubscriptionSource_1 = require("./SubscriptionSource");
const SubscriptionAction_1 = require("./SubscriptionAction");
const Phase_1 = require("./Phase");
exports.Subscription = core.serialization.object({
    id: core.serialization.string().optional(),
    locationId: core.serialization.property("location_id", core.serialization.string().optional()),
    planVariationId: core.serialization.property("plan_variation_id", core.serialization.string().optional()),
    customerId: core.serialization.property("customer_id", core.serialization.string().optional()),
    startDate: core.serialization.property("start_date", core.serialization.string().optional()),
    canceledDate: core.serialization.property("canceled_date", core.serialization.string().optionalNullable()),
    chargedThroughDate: core.serialization.property("charged_through_date", core.serialization.string().optional()),
    status: SubscriptionStatus_1.SubscriptionStatus.optional(),
    taxPercentage: core.serialization.property("tax_percentage", core.serialization.string().optionalNullable()),
    invoiceIds: core.serialization.property("invoice_ids", core.serialization.list(core.serialization.string()).optional()),
    priceOverrideMoney: core.serialization.property("price_override_money", Money_1.Money.optional()),
    version: core.serialization.bigint().optional(),
    createdAt: core.serialization.property("created_at", core.serialization.string().optional()),
    cardId: core.serialization.property("card_id", core.serialization.string().optionalNullable()),
    timezone: core.serialization.string().optional(),
    source: SubscriptionSource_1.SubscriptionSource.optional(),
    actions: core.serialization.list(SubscriptionAction_1.SubscriptionAction).optionalNullable(),
    monthlyBillingAnchorDate: core.serialization.property("monthly_billing_anchor_date", core.serialization.number().optional()),
    phases: core.serialization.list(Phase_1.Phase).optional(),
});
